tooling/metamodel-generator/src/main/java/org/hibernate/processor/ClassWriter.java

```diff
@@ -229,4 +235,24 @@ private static String writeStaticMetaModelAnnotation(Metamodel entity) {
 				: "jakarta.persistence.metamodel.StaticMetamodel";
 		return "@" + entity.importType( annotation ) + "(" + entity.getSimpleName() + ".class)";
 	}
+
+	private static void generateIdClassIfNeeded(Metamodel entity, PrintWriter pw) {
+		if ( hasAnnotation( entity.getElement(), Constants.ID_CLASS ) ) {
+			return;
+		}
+
+		final List<String> idFields = entity.getMembers().stream()
+				.filter( a -> a.inheritedAnnotations().stream()
+						.anyMatch( m -> isAnnotationMirrorOfType( m, Constants.ID ) )
+				)
+				.map( a -> entity.importType( a.getTypeDeclaration() ) + " " + a.getPropertyName() )
+				.toList();
+
+		if ( idFields.size() > 1 ) {
+			pw.print( "public record Id(" );
+			pw.print( String.join( ", ", idFields ) );
+			pw.println( ") {}" );
+			pw.println();
+		}
+	}
```

Comment:
This is nice and simple, for sure, but it's kind-of a layer-breaker. It would be better to make a new subclass of `AnnotationMeta` for this, even though I admit that's a PITA.

[I don't _love_ the architecture, but it's not terrible, and it's what we've inherited.

Reply:
You're right, but I think that generated ID class should be member of generated entity (or embeddable?) meta class. Then it should be implementing `MetaAttribute` which is less than half PITA :-)

Additional thing is that order of record components should be preserved. At least one test in core failed due to that.